In [22]:
import os, random
import keras
from keras.datasets import fashion_mnist
from keras.models import Sequential
from keras.layers import Dense, BatchNormalization, Dropout
from keras.utils import np_utils
import numpy as np
import tensorflow as tf
os.path.expanduser = lambda path: './'

In [23]:
batch_size = 128
num_classes = 10
epochs = 60

# the data, split between train and test sets
(x_train, y_train), (x_test, y_test) = fashion_mnist.load_data()

x_train = x_train.reshape(60000, 784)
x_test = x_test.reshape(10000, 784)
x_train = x_train.astype('float32')
x_test = x_test.astype('float32')
x_train /= 255
x_test /= 255
print(x_train.shape[0], 'train samples')
print(x_test.shape[0], 'test samples')

# convert class vectors to binary class matrices
y_train = keras.utils.np_utils.to_categorical(y_train, num_classes)
y_test = keras.utils.np_utils.to_categorical(y_test, num_classes)

60000 train samples
10000 test samples


# Define Model 

In [24]:
# for reproducibility
import random, os
os.environ['PYTHONHASHSEED']='0'
random.seed(123)
np.random.seed(123)
tf.random.set_seed(123)
sess = tf.compat.v1.Session(config=tf.compat.v1.ConfigProto(intra_op_parallelism_threads=1,    inter_op_parallelism_threads=1,
                                                               allow_soft_placement=True, device_count = {'CPU': 1}))
from tensorflow.python.keras import backend as K
K.set_session(sess)



kernel_initializer='glorot_uniform'
activation_function = 'relu'

with tf.device('/cpu:0'):
    model1_1 = Sequential()
    model1_1.add(Dense(512, activation='relu', input_shape=(784,)))
    model1_1.add(Dropout(0.8))
    model1_1.add(Dense(512, activation='relu'))
    model1_1.add(Dropout(0.8))
    model1_1.add(Dense(num_classes, activation='softmax'))

    model1_1.summary()

    model1_1.compile(loss='categorical_crossentropy',
                  optimizer='sgd',
                  metrics=['accuracy'])

Model: "sequential_6"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_18 (Dense)            (None, 512)               401920    
                                                                 
 dropout_12 (Dropout)        (None, 512)               0         
                                                                 
 dense_19 (Dense)            (None, 512)               262656    
                                                                 
 dropout_13 (Dropout)        (None, 512)               0         
                                                                 
 dense_20 (Dense)            (None, 10)                5130      
                                                                 
Total params: 669,706
Trainable params: 669,706
Non-trainable params: 0
_________________________________________________________________


In [25]:
    model1_2 = Sequential()
    model1_2.add(Dense(512, activation='relu', input_shape=(784,)))
    model1_2.add(Dropout(0.8))
    model1_2.add(BatchNormalization())
    model1_2.add(Dense(512, activation='relu'))
    model1_2.add(Dropout(0.8))
    model1_2.add(BatchNormalization())
    model1_2.add(Dense(num_classes, activation='softmax'))

    model1_2.summary()

    model1_2.compile(loss='categorical_crossentropy',
                  optimizer='sgd',
                  metrics=['accuracy'])

Model: "sequential_7"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_21 (Dense)            (None, 512)               401920    
                                                                 
 dropout_14 (Dropout)        (None, 512)               0         
                                                                 
 batch_normalization_6 (Batc  (None, 512)              2048      
 hNormalization)                                                 
                                                                 
 dense_22 (Dense)            (None, 512)               262656    
                                                                 
 dropout_15 (Dropout)        (None, 512)               0         
                                                                 
 batch_normalization_7 (Batc  (None, 512)              2048      
 hNormalization)                                      

In [26]:
    model2_1 = Sequential()
    model2_1.add(Dense(512, activation='relu', input_shape=(784,)))
    model2_1.add(Dropout(0.5))
    model2_1.add(Dense(512, activation='relu'))
    model2_1.add(Dropout(0.5))
    model2_1.add(Dense(num_classes, activation='softmax'))

    model2_1.summary()

    model2_1.compile(loss='categorical_crossentropy',
                  optimizer='sgd',
                  metrics=['accuracy'])

Model: "sequential_8"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_24 (Dense)            (None, 512)               401920    
                                                                 
 dropout_16 (Dropout)        (None, 512)               0         
                                                                 
 dense_25 (Dense)            (None, 512)               262656    
                                                                 
 dropout_17 (Dropout)        (None, 512)               0         
                                                                 
 dense_26 (Dense)            (None, 10)                5130      
                                                                 
Total params: 669,706
Trainable params: 669,706
Non-trainable params: 0
_________________________________________________________________


In [27]:
    model2_2 = Sequential()
    model2_2.add(Dense(512, activation='relu', input_shape=(784,)))
    model2_2.add(Dropout(0.5))
    model2_2.add(BatchNormalization())
    model2_2.add(Dense(512, activation='relu'))
    model2_2.add(Dropout(0.5))
    model2_2.add(BatchNormalization())
    model2_2.add(Dense(num_classes, activation='softmax'))

    model2_2.summary()

    model2_2.compile(loss='categorical_crossentropy',
                  optimizer='sgd',
                  metrics=['accuracy'])

Model: "sequential_9"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_27 (Dense)            (None, 512)               401920    
                                                                 
 dropout_18 (Dropout)        (None, 512)               0         
                                                                 
 batch_normalization_8 (Batc  (None, 512)              2048      
 hNormalization)                                                 
                                                                 
 dense_28 (Dense)            (None, 512)               262656    
                                                                 
 dropout_19 (Dropout)        (None, 512)               0         
                                                                 
 batch_normalization_9 (Batc  (None, 512)              2048      
 hNormalization)                                      

In [28]:
    model3_1 = Sequential()
    model3_1.add(Dense(512, activation='relu', input_shape=(784,)))
    model3_1.add(Dropout(0.2))
    model3_1.add(Dense(512, activation='relu'))
    model3_1.add(Dropout(0.2))
    model3_1.add(Dense(num_classes, activation='softmax'))

    model3_1.summary()

    model3_1.compile(loss='categorical_crossentropy',
                  optimizer='sgd',
                  metrics=['accuracy'])

Model: "sequential_10"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_30 (Dense)            (None, 512)               401920    
                                                                 
 dropout_20 (Dropout)        (None, 512)               0         
                                                                 
 dense_31 (Dense)            (None, 512)               262656    
                                                                 
 dropout_21 (Dropout)        (None, 512)               0         
                                                                 
 dense_32 (Dense)            (None, 10)                5130      
                                                                 
Total params: 669,706
Trainable params: 669,706
Non-trainable params: 0
_________________________________________________________________


In [29]:
    model3_2 = Sequential()
    model3_2.add(Dense(512, activation='relu', input_shape=(784,)))
    model3_2.add(Dropout(0.2))
    model3_2.add(BatchNormalization())
    model3_2.add(Dense(512, activation='relu'))
    model3_2.add(Dropout(0.2))
    model3_2.add(BatchNormalization())
    model3_2.add(Dense(num_classes, activation='softmax'))

    model3_2.summary()

    model3_2.compile(loss='categorical_crossentropy',
                  optimizer='sgd',
                  metrics=['accuracy'])

Model: "sequential_11"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_33 (Dense)            (None, 512)               401920    
                                                                 
 dropout_22 (Dropout)        (None, 512)               0         
                                                                 
 batch_normalization_10 (Bat  (None, 512)              2048      
 chNormalization)                                                
                                                                 
 dense_34 (Dense)            (None, 512)               262656    
                                                                 
 dropout_23 (Dropout)        (None, 512)               0         
                                                                 
 batch_normalization_11 (Bat  (None, 512)              2048      
 chNormalization)                                    

# Start Training

In [30]:
history1_1 = model1_1.fit(x_train, y_train,
                    batch_size=batch_size,
                    epochs=epochs,
                    validation_split=0.2)

Epoch 1/60
375/375 [==============================] - 3s 9ms/step - loss: 1.9859 - accuracy: 0.2970 - val_loss: 1.2499 - val_accuracy: 0.6512
Epoch 2/60
375/375 [==============================] - 3s 8ms/step - loss: 1.3897 - accuracy: 0.4929 - val_loss: 0.9018 - val_accuracy: 0.6895
Epoch 3/60
375/375 [==============================] - 3s 8ms/step - loss: 1.1611 - accuracy: 0.5704 - val_loss: 0.7840 - val_accuracy: 0.7170
Epoch 4/60
375/375 [==============================] - 3s 8ms/step - loss: 1.0343 - accuracy: 0.6180 - val_loss: 0.7241 - val_accuracy: 0.7377
Epoch 5/60
375/375 [==============================] - 3s 8ms/step - loss: 0.9531 - accuracy: 0.6456 - val_loss: 0.6814 - val_accuracy: 0.7557
Epoch 6/60
375/375 [==============================] - 3s 8ms/step - loss: 0.8955 - accuracy: 0.6672 - val_loss: 0.6520 - val_accuracy: 0.7623
Epoch 7/60
375/375 [==============================] - 3s 8ms/step - loss: 0.8535 - accuracy: 0.6860 - val_loss: 0.6265 - val_accuracy: 0.7732
Epoch 

In [31]:
history1_2 = model1_2.fit(x_train, y_train,
                    batch_size=batch_size,
                    epochs=epochs,
                    validation_split=0.2)

Epoch 1/60
375/375 [==============================] - 4s 11ms/step - loss: 2.0238 - accuracy: 0.3215 - val_loss: 1.0047 - val_accuracy: 0.7088
Epoch 2/60
375/375 [==============================] - 4s 10ms/step - loss: 1.2973 - accuracy: 0.5310 - val_loss: 0.7790 - val_accuracy: 0.7422
Epoch 3/60
375/375 [==============================] - 4s 10ms/step - loss: 1.0758 - accuracy: 0.6057 - val_loss: 0.7130 - val_accuracy: 0.7467
Epoch 4/60
375/375 [==============================] - 4s 10ms/step - loss: 0.9667 - accuracy: 0.6446 - val_loss: 0.6667 - val_accuracy: 0.7635
Epoch 5/60
375/375 [==============================] - 4s 10ms/step - loss: 0.8976 - accuracy: 0.6689 - val_loss: 0.6453 - val_accuracy: 0.7741
Epoch 6/60
375/375 [==============================] - 4s 10ms/step - loss: 0.8478 - accuracy: 0.6900 - val_loss: 0.6196 - val_accuracy: 0.7791
Epoch 7/60
375/375 [==============================] - 4s 10ms/step - loss: 0.8142 - accuracy: 0.7042 - val_loss: 0.6025 - val_accuracy: 0.7862

Epoch 58/60
375/375 [==============================] - 4s 10ms/step - loss: 0.5082 - accuracy: 0.8226 - val_loss: 0.4072 - val_accuracy: 0.8558
Epoch 59/60
375/375 [==============================] - 4s 10ms/step - loss: 0.5007 - accuracy: 0.8267 - val_loss: 0.4084 - val_accuracy: 0.8583
Epoch 60/60
375/375 [==============================] - 4s 11ms/step - loss: 0.4987 - accuracy: 0.8251 - val_loss: 0.4069 - val_accuracy: 0.8568


In [32]:
history2_1 = model2_1.fit(x_train, y_train,
                    batch_size=batch_size,
                    epochs=epochs,
                    validation_split=0.2)

Epoch 1/60
375/375 [==============================] - 4s 10ms/step - loss: 1.4087 - accuracy: 0.5072 - val_loss: 0.8207 - val_accuracy: 0.7198
Epoch 2/60
375/375 [==============================] - 4s 9ms/step - loss: 0.9034 - accuracy: 0.6818 - val_loss: 0.6809 - val_accuracy: 0.7568
Epoch 3/60
375/375 [==============================] - 4s 10ms/step - loss: 0.7810 - accuracy: 0.7254 - val_loss: 0.6170 - val_accuracy: 0.7846
Epoch 4/60
375/375 [==============================] - 4s 9ms/step - loss: 0.7135 - accuracy: 0.7484 - val_loss: 0.5797 - val_accuracy: 0.7988
Epoch 5/60
375/375 [==============================] - 4s 10ms/step - loss: 0.6665 - accuracy: 0.7651 - val_loss: 0.5459 - val_accuracy: 0.8106
Epoch 6/60
375/375 [==============================] - 4s 10ms/step - loss: 0.6306 - accuracy: 0.7826 - val_loss: 0.5266 - val_accuracy: 0.8142
Epoch 7/60
375/375 [==============================] - 3s 9ms/step - loss: 0.6044 - accuracy: 0.7904 - val_loss: 0.5061 - val_accuracy: 0.8233
Ep

In [33]:
history2_2 = model2_2.fit(x_train, y_train,
                    batch_size=batch_size,
                    epochs=epochs,
                    validation_split=0.2)

Epoch 1/60
375/375 [==============================] - 4s 10ms/step - loss: 1.0762 - accuracy: 0.6294 - val_loss: 0.5882 - val_accuracy: 0.7991
Epoch 2/60
375/375 [==============================] - 4s 10ms/step - loss: 0.6925 - accuracy: 0.7548 - val_loss: 0.4814 - val_accuracy: 0.8256
Epoch 3/60
375/375 [==============================] - 4s 10ms/step - loss: 0.6116 - accuracy: 0.7830 - val_loss: 0.4483 - val_accuracy: 0.8373
Epoch 4/60
375/375 [==============================] - 4s 10ms/step - loss: 0.5706 - accuracy: 0.7973 - val_loss: 0.4313 - val_accuracy: 0.8438
Epoch 5/60
375/375 [==============================] - 4s 10ms/step - loss: 0.5374 - accuracy: 0.8089 - val_loss: 0.4157 - val_accuracy: 0.8487
Epoch 6/60
375/375 [==============================] - 4s 10ms/step - loss: 0.5141 - accuracy: 0.8160 - val_loss: 0.4093 - val_accuracy: 0.8513
Epoch 7/60
375/375 [==============================] - 4s 10ms/step - loss: 0.4933 - accuracy: 0.8224 - val_loss: 0.3987 - val_accuracy: 0.8548

Epoch 58/60
375/375 [==============================] - 4s 10ms/step - loss: 0.2947 - accuracy: 0.8918 - val_loss: 0.3065 - val_accuracy: 0.8884
Epoch 59/60
375/375 [==============================] - 4s 10ms/step - loss: 0.2897 - accuracy: 0.8928 - val_loss: 0.3056 - val_accuracy: 0.8899
Epoch 60/60
375/375 [==============================] - 4s 10ms/step - loss: 0.2956 - accuracy: 0.8913 - val_loss: 0.3050 - val_accuracy: 0.8907


In [34]:
history3_1 = model3_1.fit(x_train, y_train,
                    batch_size=batch_size,
                    epochs=epochs,
                    validation_split=0.2)

Epoch 1/60
375/375 [==============================] - 4s 9ms/step - loss: 1.2041 - accuracy: 0.6135 - val_loss: 0.7423 - val_accuracy: 0.7603
Epoch 2/60
375/375 [==============================] - 3s 9ms/step - loss: 0.7479 - accuracy: 0.7492 - val_loss: 0.6205 - val_accuracy: 0.7883
Epoch 3/60
375/375 [==============================] - 3s 9ms/step - loss: 0.6475 - accuracy: 0.7823 - val_loss: 0.5593 - val_accuracy: 0.8113
Epoch 4/60
375/375 [==============================] - 3s 9ms/step - loss: 0.5976 - accuracy: 0.7968 - val_loss: 0.5264 - val_accuracy: 0.8215
Epoch 5/60
375/375 [==============================] - 3s 9ms/step - loss: 0.5621 - accuracy: 0.8081 - val_loss: 0.5052 - val_accuracy: 0.8255
Epoch 6/60
375/375 [==============================] - 3s 9ms/step - loss: 0.5373 - accuracy: 0.8160 - val_loss: 0.4862 - val_accuracy: 0.8319
Epoch 7/60
375/375 [==============================] - 4s 9ms/step - loss: 0.5168 - accuracy: 0.8224 - val_loss: 0.4721 - val_accuracy: 0.8383
Epoch 

In [35]:
history3_2 = model3_2.fit(x_train, y_train,
                    batch_size=batch_size,
                    epochs=epochs,
                    validation_split=0.2)

Epoch 1/60
375/375 [==============================] - 4s 10ms/step - loss: 0.7353 - accuracy: 0.7451 - val_loss: 0.5036 - val_accuracy: 0.8258
Epoch 2/60
375/375 [==============================] - 4s 10ms/step - loss: 0.5155 - accuracy: 0.8184 - val_loss: 0.4163 - val_accuracy: 0.8515
Epoch 3/60
375/375 [==============================] - 4s 10ms/step - loss: 0.4580 - accuracy: 0.8363 - val_loss: 0.3880 - val_accuracy: 0.8602
Epoch 4/60
375/375 [==============================] - 4s 10ms/step - loss: 0.4308 - accuracy: 0.8462 - val_loss: 0.3751 - val_accuracy: 0.8637
Epoch 5/60
375/375 [==============================] - 4s 10ms/step - loss: 0.4072 - accuracy: 0.8543 - val_loss: 0.3619 - val_accuracy: 0.8707
Epoch 6/60
375/375 [==============================] - 4s 10ms/step - loss: 0.3905 - accuracy: 0.8594 - val_loss: 0.3583 - val_accuracy: 0.8696
Epoch 7/60
375/375 [==============================] - 4s 10ms/step - loss: 0.3746 - accuracy: 0.8643 - val_loss: 0.3504 - val_accuracy: 0.8733

Epoch 58/60
375/375 [==============================] - 4s 10ms/step - loss: 0.1623 - accuracy: 0.9409 - val_loss: 0.3113 - val_accuracy: 0.8972
Epoch 59/60
375/375 [==============================] - 4s 10ms/step - loss: 0.1557 - accuracy: 0.9434 - val_loss: 0.3153 - val_accuracy: 0.8949
Epoch 60/60
375/375 [==============================] - 4s 10ms/step - loss: 0.1567 - accuracy: 0.9417 - val_loss: 0.3231 - val_accuracy: 0.8931


# Calculate accuracy 

In [36]:
metrics1_1 = model1_1.evaluate(x_test, y_test) #returns loss and accuracy
print("Dropout Rate : 0.8")
print(metrics1_1[1])
print(f'Accuracy: {metrics1_1[1]*100:.2f}%\n')

313/313 [==============================] - 0s 1ms/step - loss: 0.4215 - accuracy: 0.8476
Dropout Rate : 0.8
0.847599983215332
Accuracy: 84.76%



In [37]:
metrics1_2 = model1_2.evaluate(x_test, y_test) #returns loss and accuracy
print("Dropout Rate : 0.8 with batchnormalization")
print(metrics1_2[1])
print(f'Accuracy: {metrics1_2[1]*100:.2f}%\n')

313/313 [==============================] - 0s 1ms/step - loss: 0.4293 - accuracy: 0.8489
Dropout Rate : 0.8 with batchnormalization
0.8489000201225281
Accuracy: 84.89%



In [38]:
metrics2_1 = model2_1.evaluate(x_test, y_test) #returns loss and accuracy
print("Dropout Rate : 0.5")
print(metrics2_1[1])
print(f'Accuracy: {metrics2_1[1]*100:.2f}%\n')

313/313 [==============================] - 0s 1ms/step - loss: 0.3624 - accuracy: 0.8701
Dropout Rate : 0.5
0.8701000213623047
Accuracy: 87.01%



In [39]:
metrics2_2 = model2_2.evaluate(x_test, y_test) #returns loss and accuracy
print("Dropout Rate : 0.5 with batchnormalization")
print(metrics2_2[1])
print(f'Accuracy: {metrics2_2[1]*100:.2f}%\n')

313/313 [==============================] - 0s 2ms/step - loss: 0.3262 - accuracy: 0.8848
Dropout Rate : 0.5 with batchnormalization
0.8848000168800354
Accuracy: 88.48%



In [40]:
metrics3_1 = model3_1.evaluate(x_test, y_test) #returns loss and accuracy
print("Dropout Rate : 0.2")
print(metrics3_1[1])
print(f'Accuracy: {metrics3_1[1]*100:.2f}%\n')

313/313 [==============================] - 0s 1ms/step - loss: 0.3481 - accuracy: 0.8767
Dropout Rate : 0.2
0.8766999840736389
Accuracy: 87.67%



In [41]:
metrics3_2 = model3_2.evaluate(x_test, y_test) #returns loss and accuracy
print("Dropout Rate : 0.2 with batchnormalization")
print(metrics3_2[1])
print(f'Accuracy: {metrics3_2[1]*100:.2f}%\n')

313/313 [==============================] - 0s 1ms/step - loss: 0.3431 - accuracy: 0.8899
Dropout Rate : 0.2 with batchnormalization
0.8899000287055969
Accuracy: 88.99%



# 분석결과

저는 dropout rate를 0.8, 0.5, 0.2순으로 학습을 수행해보았습니다.

다음은 그 결과로 나온 정확도입니다.

-Dropout Rate 0.2    / with BatchNormalization

Accuracy: 87.67%     / Accuracy: 88.99%

-Dropout Rate 0.5    / with BatchNormalization

Accuracy: 87.01%     / Accuracy: 88.48%

-Dropout Rate 0.8    / with BatchNormalization

Accuracy: 84.76%     / Accuracy: 84.89%

Dropout Rate이 각각 0.2, 0.5, 0.8로 dropout rate이 증가할수록 정확도는

점점 감소하는 것을 확인할 수 있다. 이는 drop하는 neuron의 수가 많아질수록

정확도는 떨어지지만 overfitting은 피할 수 있다는 것을 확인할 수 있다.

각각의 Dropout에서 Batch Normalization을 적용한 경우 dropout rate가 0.2, 0.5, 0.8인 경우

정확도가 조금 증가한 것을 볼 수 있는데 이는 batch normalization과 dropout이 비슷한 역할을 하기 때문에

정확도가 조금 증가한 것으로 볼 수 있다.